File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int i: nums) sum+=i;

        if(sum%2!=0) return false;

        HashSet<Integer> dp = new HashSet<>();
        dp.add(nums[0]);

        for(int i=1; i<nums.length; i++){
            HashSet<Integer> newDP = new HashSet<>();
            for(Integer j : dp){
                int newSum = nums[i] + j;
                if(newSum == sum/2) return true;
                newDP.add(newSum);
                newDP.add(j);
            }
            dp = newDP;
        }

        return dp.contains(sum/2);
    }
}

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.py
class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        n = len(nums)
        total = sum(nums)

        def helper(idx, sumSoFar, memo):
            if idx == n:
                return False

            if (idx, sumSoFar) in memo:
                return memo[(idx, sumSoFar)]

            if sumSoFar == total - sumSoFar:
                return True

            #Pick 
            if helper(idx+1, sumSoFar+nums[idx], memo):
                memo[(idx, sumSoFar)] = True
                return True

            #Not Pick
            if helper(idx+1, sumSoFar, memo):
                memo[(idx, sumSoFar)] = True
                return True
            
            memo[(idx, sumSoFar)] = False
            return False


        memo = {}
        return helper(0, 0, memo)

        total = sum(nums)
        if total % 2 != 0:
            return False

        target = total//2
        dp = [False] * (target + 1)
        dp[0] = True #Base case

        for num in nums:
            for j in range(target, num - 1, -1):
                dp[j] = dp[j] or dp[j - num]

        return dp[target]

       

        